File: binding/include/webchannel/libqwebchannelabstracttransport.zig
const QtC = @import("qt6zig");
const qtc = @import("qt6c");%%_IMPORTLIBS_%%
%%_STRUCTDEFS_%%

/// https://doc.qt.io/qt-6/qwebchannelabstracttransport.html
pub const qwebchannelabstracttransport = struct {

    /// New constructs a new QWebChannelAbstractTransport object.
    pub fn New() QtC.QWebChannelAbstractTransport {
        return qtc.QWebChannelAbstractTransport_new();
    }


    /// New2 constructs a new QWebChannelAbstractTransport object.
    pub fn New2(parent: ?*anyopaque) QtC.QWebChannelAbstractTransport {
        return qtc.QWebChannelAbstractTransport_new2(@ptrCast(parent));
    }


    pub fn Tr(s: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebChannelAbstractTransport_Tr(@ptrCast(s));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebchannelabstracttransport.Tr: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn SendMessage(self: ?*anyopaque, message: ?*anyopaque) void {
        qtc.QWebChannelAbstractTransport_SendMessage(@ptrCast(self), @ptrCast(message));
    }

    /// Allows for overriding the related default method
    pub fn OnSendMessage(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QWebChannelAbstractTransport_OnSendMessage(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    /// Base class method implementation
    pub fn QBaseSendMessage(self: ?*anyopaque, message: ?*anyopaque) void {
        qtc.QWebChannelAbstractTransport_QBaseSendMessage(@ptrCast(self), @ptrCast(message));
    }

    pub fn MessageReceived(self: ?*anyopaque, message: ?*anyopaque, transport: ?*anyopaque) void {
        qtc.QWebChannelAbstractTransport_MessageReceived(@ptrCast(self), @ptrCast(message), @ptrCast(transport));
    }

    pub fn OnMessageReceived(self: ?*anyopaque, callback: *const fn (?*anyopaque, ?*anyopaque, ?*anyopaque) callconv(.c) void) void {
        qtc.QWebChannelAbstractTransport_Connect_MessageReceived(@ptrCast(self), @intCast(@intFromPtr(callback)));
    }

    pub fn Tr2(s: []const u8, c: []const u8, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebChannelAbstractTransport_Tr2(@ptrCast(s), @ptrCast(c));
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebchannelabstracttransport.Tr2: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    pub fn Tr3(s: []const u8, c: []const u8, n: i32, allocator: std.mem.Allocator) []const u8 {
        const _str = qtc.QWebChannelAbstractTransport_Tr3(@ptrCast(s), @ptrCast(c), n);
defer qtc.libqt_string_free(&_str);
const _ret = allocator.alloc(u8, _str.len) catch @panic("qwebchannelabstracttransport.Tr3: Memory allocation failed");
@memcpy(_ret, _str.data[0.._str.len]);
return  _ret;
    }

    /// Delete this object from C++ memory.
    pub fn QDelete(self: ?*anyopaque) void {
        qtc.QWebChannelAbstractTransport_Delete(@ptrCast(self));
    }
};

File: binding/src/webchannel/libqwebchannelabstracttransport.cpp
#include <QJsonObject>
#include <QObject>
#include <QString>
#include <QByteArray>
#include <cstring>
#include <type_traits>
#include <QWebChannelAbstractTransport>
#include <qwebchannelabstracttransport.h>
#include "libqwebchannelabstracttransport.h"
#include "libqwebchannelabstracttransport.hxx"

QWebChannelAbstractTransport* QWebChannelAbstractTransport_new() {
	 return new VirtualQWebChannelAbstractTransport();
}

QWebChannelAbstractTransport* QWebChannelAbstractTransport_new2(QObject* parent) {
	 return new VirtualQWebChannelAbstractTransport(parent);
}

libqt_string QWebChannelAbstractTransport_Tr(const char* s) {
	QString _ret = QWebChannelAbstractTransport::tr(s);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

void QWebChannelAbstractTransport_SendMessage(QWebChannelAbstractTransport* self, const QJsonObject* message) {
	self->sendMessage(*message);
}

void QWebChannelAbstractTransport_MessageReceived(QWebChannelAbstractTransport* self, const QJsonObject* message, QWebChannelAbstractTransport* transport) {
	self->messageReceived(*message, transport);
}

void QWebChannelAbstractTransport_Connect_MessageReceived(QWebChannelAbstractTransport* self, intptr_t slot) {
    void (*slotFunc)(QWebChannelAbstractTransport*, const QJsonObject*, QWebChannelAbstractTransport*) = reinterpret_cast<void (*)(QWebChannelAbstractTransport*, const QJsonObject*, QWebChannelAbstractTransport*)>(slot);
    QWebChannelAbstractTransport::connect(self, &QWebChannelAbstractTransport::messageReceived, [self, slotFunc](const QJsonObject& message, QWebChannelAbstractTransport* transport) {
	slotFunc(self, message, transport);
    });
}

libqt_string QWebChannelAbstractTransport_Tr2(const char* s, const char* c) {
	QString _ret = QWebChannelAbstractTransport::tr(s, c);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

libqt_string QWebChannelAbstractTransport_Tr3(const char* s, const char* c, int n) {
	QString _ret = QWebChannelAbstractTransport::tr(s, c, n);
	QByteArray _b = _ret.toUtf8();
	libqt_string _str;
	_str.len = _b.length();
	_str.data = static_cast<const char*>(malloc(_str.len + 1));
	memcpy((void*)_str.data, _b.data(), _str.len);
	((char*)_str.data)[_str.len] = '\0';
	return _str;
}

// Base class handler implementation
void QWebChannelAbstractTransport_QBaseSendMessage(QWebChannelAbstractTransport* self, const QJsonObject* message) {
	auto* vqwebchannelabstracttransport = dynamic_cast<VirtualQWebChannelAbstractTransport*>(self);
	if (vqwebchannelabstracttransport && vqwebchannelabstracttransport->isVirtualQWebChannelAbstractTransport) {
vqwebchannelabstracttransport->setQWebChannelAbstractTransport_SendMessage_IsBase(true);
	vqwebchannelabstracttransport->sendMessage(*message);
}
}

// Auxiliary method to allow providing re-implementation
void QWebChannelAbstractTransport_OnSendMessage(QWebChannelAbstractTransport* self, intptr_t slot) {
	auto* vqwebchannelabstracttransport = dynamic_cast<VirtualQWebChannelAbstractTransport*>(self);
	if (vqwebchannelabstracttransport && vqwebchannelabstracttransport->isVirtualQWebChannelAbstractTransport) {
vqwebchannelabstracttransport->setQWebChannelAbstractTransport_SendMessage_Callback(reinterpret_cast<VirtualQWebChannelAbstractTransport::QWebChannelAbstractTransport_SendMessage_Callback>(slot));
}
}

void QWebChannelAbstractTransport_Delete(QWebChannelAbstractTransport* self) {
    delete self;
}

